### ISSUE TITLE
Review Chương 1 (Lần 2)

### ISSUE THREAD

**Issue by KienTrann** [hoangminhquan-lhsdt/optimizers#5]
## Một số ý ở mức nói chung

- Về mặt thuật ngữ, mình thấy tụi em dùng có vẻ hơi không chuẩn. Cụ thể, "đạo hàm" là nói về hàm một biến; ở đây, hàm của mình là hàm nhiều biến nên chỉ có "đạo hàm riêng". Còn gradient = véc-tơ các đạo hàm riêng. Lần đầu tiên nhắc đến thuật ngữ "gradient" (hay một thuật ngữ nào khác) thì tụi em nên có giải thích qua để người đọc nắm được.
- Có vẻ tụi em chọn trước trích dẫn [...] không có khoảng trắng. Nhưng với hầu hết các paper mình đã đọc thì mình nhớ là có khoảng trắng trước [...] (ví dụ như [paper này](https://arxiv.org/pdf/1802.05814.pdf)). 

### Đoạn văn "Việc tối ưu hoá mạng nơ-ron có thể hiểu là ..."

> mặt phẳng lỗi

Mình nghĩ dịch "error surface" là "mặt phẳng lỗi" thì có lẽ không hợp lý lắm vì có "phẳng" không?

> sự hỗn loạn của mặt phẳng lỗi

Mình nghĩ thay "hỗn loạn" bằng "phức tạp" thì sẽ tốt hơn

> ... cấu thành từ nhiều “critical point” (Hình 1.3), những điểm có gradient bằng 0, như cực tiểu địa phương, điểm yên ngựa, hợp với các vùng bằng phẳng, và
vùng rãnh hẹp

Gợi ý sửa: \
"... cấu thành từ nhiều “critical point” ("critical point" là những điểm có gradient bằng 0 như cực tiểu địa phương, điểm yên ngựa) hợp với các vùng bằng phẳng, và vùng rãnh hẹp (Hình 1.3)" 

### Đoạn văn "Các điểm cực tiểu địa phương từng được xem ..."

> Yann N. Dauphin cũng cho rằng khi số lượng tầng ẩn quá lớn thì tất cả các điểm critical point sẽ là điểm yên ngựa

**Tất cả** critical point đều là điểm yên ngựa? Vậy là không có điểm thuần cực tiểu?  

> Vì thế, việc tránh và tìm được cách thoát ra khỏi điểm yên ngựa ...

Mình nghĩ bỏ câu này đi vì câu này không kết nối tốt với câu trước đó. Câu này cũng không cần thiết trong đoạn văn này

### Đoạn văn "Ngoài các điểm yên ngựa, bề mặt lỗi còn ..."

> Hiện tượng này [rãnh hẹp] xuất hiện thường xuyên trong mạng nơ-ron nhiều tầng ẩn do ảnh hưởng của mỗi trọng số lên độ lỗi là khác nhau và giá trị trọng số sẽ phụ thuộc vào giá trị trọng số ở các tầng trước đó.

Mình thấy lý do đưa ra không rõ ràng lắm. "ảnh hưởng của mỗi trọng số lên độ lỗi là khác nhau": mạng nơ-ron nông cũng vậy mà. "giá trị trọng số sẽ phụ thuộc vào giá trị trọng số ở các tầng trước đó": mình đọc mà không hiểu gì lắm.

> Một giải pháp thường được sử dụng để vượt qua các vùng này là lựa chọn một bước cập nhật đủ nhỏ để tránh dao động quanh hướng có độ cong cao nhưng vẫn đủ để tiếp tục di chuyển theo hướng có độ cong thấp về hướng cho độ lỗi thấp hơn. Tuy nhiên, cách giải quyết này cho thời gian huấn luyện quá chậm để có ý nghĩa thực tiễn.

Gợi ý sửa: bỏ phần này đi vì ý của đoạn văn này và 2 đoạn văn trước là mô tả về sự phức tạp của bề mặt lỗi, còn các phương pháp để giải quyết là ở phần sau, tụi em không nên nhét vào ở đây (mình thấy ở phần sau tụi em cũng đã có nói về vụ dùng learning rate nhỏ này).

## Phần về các phương pháp giải quyết bài toán mà người ta đã làm

Trong phần này, đáng lẽ ra phải có trích dẫn về các phương pháp mà người ta đã làm nhưng mình thấy tụi em chỉ mới có trích dẫn Adam.

Mình có tóm tắt lại ý của các đoạn văn mà tụi em viết về phần này (mỗi chấm tròn ứng với một đoạn văn):

- Phương pháp bậc nhất ít bị mắc kẹt tại điểm yên ngựa, nhưng dễ bị mắc kẹt ở các vùng "bằng phẳng" 
- Phương pháp bậc nhất tính gradient chậm do phải duyệt qua toàn bộ tập dữ liệu --> dùng một tập con để bước lượng gradient 
- Phương pháp dùng một tập con vẫn gặp trở ngại tại các cực tiểu địa phương, vùng bằng phẳng, rãnh hẹp --> dùng momentum
- Phương pháp momentum có ưu điểm ...
- Phương pháp momentum có nhược điểm là phải chọn thêm một siêu tham số
- Phương pháp bậc nhất hay xảy ra dao động do hướng của gradient không trùng với hướng để đi tới cực tiểu --> phương pháp bậc hai. Phương pháp bậc hai có nhược điểm là dễ bị mắc kẹt tại điểm yên ngựa và tốn chi phí để tính ma trận Hessian --> xấp xỉ ma trận Hessian (phương pháp tỉ lệ học thích ứng)
- Phương pháp tỉ lệ học thích ứng có ưu điểm ...
- Phương pháp tỉ lệ học thích ứng có nhược điểm là dao động tại vùng gần cực tiểu và không chạy tốt khi các hướng low/high curvature không align với các trục trọng số 

Nói chung, mình thấy tụi em viết còn dài dòng; cần viết súc tích lại, tập trung vào big idea của mỗi phương pháp, tránh nói quá chi tiết. Ý của các đoạn văn cũng chưa liên kết tốt với nhau. Ví dụ: phương pháp bậc nhất (GD) thì tụi em chỉ nói nhược điểm là dễ mắc kẹt ở vùng bằng phẳng, còn phương pháp dùng tập con (SGD) thì tụi em nói là dễ mắc kẹt ở cực tiểu địa phương, vùng bằng phẳng, rãnh hẹp; phương pháp momentum thì tụi em nói về nhược điểm là phải chọn thêm siêu tham số nhưng phương pháp tỉ lệ học thích ứng thì không thấy tụi em nói gì (ngoài ra, cũng nên xem xét là có nên nói về vụ phải chọn siêu tham số không? vụ này có vẻ là ở mức chi tiết, không phải ở mức big idea); ở chỗ chuyển tiếp qua phương pháp bậc hai, tụi em nói là phương pháp bậc nhất hay xảy ra dao động do ..., "phương pháp bậc nhất" mà tụi em dùng có bao gồm momentum (nếu bao gồm thì ý tụi em nói không đúng, nếu không bao gồm thì không kết nối được với đoạn về momentum ở trước đó)

Mình gợi ý sửa cấu trúc của phần này như sau:

- Nói idea của GD (nói idea thôi, không phân tích gì thêm), rồi nói về việc tính gradient trên toàn bộ tập dữ liệu huấn luyện sẽ chậm khi tập dữ liệu lớn (setting thường thấy khi huấn luyện mạng nơ-ron nhiều tầng ẩn) nên trong thực tế người ta thường không dùng GD mà dùng SGD (nói về idea của SGD). Từ đây về sau, chỉ tập trung vào nhân vật chính là SGD (mặc dù khi cần thì cũng có thể liên hệ với GD một xíu). Lúc này, có thể nói về ưu điểm của SGD là ít bị kẹt ở điểm yên ngựa. 
- Nói về vấn đề của SGD khi gặp vùng có độ nghiêng nhỏ và vùng rãnh hẹp. Từ đó nói về idea của momentum để giúp khắc phục vấn đề gặp phải của SGD (ngoài khắc phục vấn đề khi gặp vùng có độ nghiêng nhỏ và vùng rãnh hẹp thì momentum cũng phần nào giúp giảm bớt sự stochastic của SGD). [Ở đây, mình không biết là có nên nói về vấn đề của SGD khi gặp cực tiểu cục bộ không, nếu nói thì đó là vấn đề gì, liệu có mâu thuẫn với ý cực tiểu cục bộ không phải là vấn đề mà tụi em đã nói ở trên không, momentum có giúp khắc phục vấn đề này không]
- Nói về vấn đề của SGD + momentum với bề mặt lỗi có phổ độ lớn của các đạo hàm riêng **rất** khác nhau. Từ đó, nói về idea của phương pháp dùng đạo hàm riêng bậc hai để khắc phục vấn đề gặp phải của SGD + momentum
- Nói về vấn đề chi phí tính toán lớn của phương pháp dùng đạo hàm riêng bậc hai. Từ đó nói về idea của phương pháp tỉ lệ học thích ứng để giải quyết vấn đề của phương pháp dùng đạo hàm riêng bậc 2. Cũng nói thêm: do chỉ xấp xỉ ma trận Hessian bằng cách tính (tính xấp xỉ?) phần đường chéo của ma trận Hessian nên sẽ không chạy tốt khi các hướng low/high curvature của bề mặt lỗi không align với các trục trọng số (ở đây, cũng nên liên hệ với momentum). [Ở đây, mình không biết có nên nói về vấn đề phương pháp bậc 2 dễ bị mắc vào điểm yên ngựa không, nếu vậy thì phương pháp tỉ lệ học thích ứng có bị không]   

> Thuật toán Adam sử dụng trung bình chạy để tích tụ quán tính kết hợp với phương pháp xấp xỉ ma trận Hessian của phương pháp tỉ lệ học thích ứng cho phép Adam hoạt động tốt trong trường hợp dữ liệu thưa

Cần giải thích về "dữ liệu thưa" để kết nối vào những ý đã nói ở các đoạn văn bên trên.

> Nhờ sự kết hợp này, Adam kế thừa được nhiều ưu điểm, đồng thời khắc phục những hạn chế của các hướng tiếp cận trên khi có thể tăng giảm độ dài bước cập nhật tùy theo điều kiện bề mặt lỗi.

Câu này không rõ (Gradient Descent cũng "tăng giảm độ dài bước cập nhật" được mà). Mình nghĩ có thể bỏ câu này đi.

**Comment by KienTrann** [comment]
Mình nói thêm một xíu cho rõ.

> Nói về vấn đề của SGD + momentum với bề mặt lỗi có phổ độ lớn của các đạo hàm riêng rất khác nhau.

Ý của mình là ám chỉ vùng "rãnh hẹp", nhưng có các chiều có curvature **rất** nhỏ và có các chiều có curvature **rất** lớn. Vụ "dữ liệu thưa" (hay "gradient thưa") mà tụi em có nói chắc cũng thuộc trường hợp này.
